daml_dit_if/main/jwt.py

```diff
@@ -0,0 +1,148 @@
+from asyncio import shield, sleep
+from datetime import timedelta
+import json
+from typing import TYPE_CHECKING, Any, Collection, Mapping, Optional, Union
+
+from aiohttp import ClientSession
+
+if TYPE_CHECKING:
+    from jwcrypto.jwk import JWK
+
+from .log import LOG
+
+
+
+DEFAULT_POLL_INTERVAL = timedelta(seconds=10)
+IAT_SKEW = timedelta(seconds=15)
+MAX_TOKEN_EXPIRY = timedelta(days=1)
+
+
+class JWTValidator:
+    def __init__(self, jwks_urls: "Optional[Union[str, Collection[str]]]" = None):
+        from jwcrypto.jwk import JWKSet
+
+        self.jwks_urls = jwks_urls
+        self.keys = JWKSet()
+        self.session = None  # type: Optional[ClientSession]
+
+    async def poll(self, poll_interval: timedelta = DEFAULT_POLL_INTERVAL):
+        """
+        Periodically check for new keys. This coroutine will NEVER terminate naturally, so it should
+        not be awaited.
+        """
+        while True:
+            await sleep(poll_interval.total_seconds())
+
+            LOG.debug("JWKS poller polling for new keys...")
+            await shield(self._load_new_keys())
+            LOG.debug("...JWKS poll complete.")
+
+    async def decode_claims(self, token: str) -> "Mapping[str, Any]":
+        from jwcrypto.jwt import JWT
+
+        LOG.debug("Verifying token: %r", token)
+        jwt = JWT(jwt=token)
+        key = jwt.token.jose_header["kid"]
+        await self.get_key(key)
+
+        jwt = JWT(jwt=token, key=self.keys)
+        return json.loads(jwt.claims)
+
+    async def get_key(self, kid: str) -> "Optional[JWK]":
+        """
+        Retrieve a key for a given ``kid``. If the key could not be found, the keystore is
+        refreshed, and if a key is discovered through that process, it is returned. May return
+        ``None`` if even after refreshing keys from the remote JWKS source, a key for the given
+        ``kid`` could not be found.
+        """
+        key = self.keys.get_key(kid)
+        if key is None:
+            await shield(self._load_new_keys())
+            key = self.keys.get_key(kid)
+
+        return key
+
+    def export_all_keys(self) -> "str":
+        """
+        Return a JSON string that formats all public keys currently in the store as JWKS.
+        """
+        return self.keys.export(private_keys=False)
+
+    async def _load_new_keys(self):
+        from jwcrypto.jwk import JWK
+
+        for url in self.jwks_urls:
+            try:
+                if self.session is None:
+                    self.session = ClientSession()
+
+                async with self.session.get(url, allow_redirects=False) as response:
+                    jwks_json = await response.json()
+
+                # ``JWKSet.import_keyset`` suffers from a few critical flaws that make it
```

Comment:
ugh…it lives again…

Reply:
oh wait actually

So this is a copy of some code from `dabl_common`? Perhaps this should be pulled out more properly; if there's a better place to put it in one of the existing OSS repos then maybe we move it.

Reply:
> Not a fan of the JWT copy-paste

Neither am I.

>  as it's tricky and annoying code, 

It is.

> but I won't make that a blocking concern.

Thanks... I do think this goes on the list of stuff to fix.